cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTable.java

```diff
@@ -278,33 +278,33 @@ public CloseableIterator<StructuredRow> multiScan(Collection<Range> keyRanges,
    * clause using the {@link #appendRange(StringBuilder, Range)} method. The where clause of each range are OR together.
    * E.g.
    * <p>
-   * SELECT * FROM table WHERE key1 in (?,?) AND key2 in (?,?)
+   * SELECT * FROM table WHERE (key1 = ? AND key2 = ?) OR (key1 = ? AND key2 = ?)
    * OR ((key3 >= ?) AND (key3 <= ?)) OR ((key4 >= ?) AND (key4 <= ?)) LIMIT limit
    *
-   * @param keyFields a map from field name to field values that the query has to match with
+   * @param singletonRanges the list of singleton ranges to scan
    * @param ranges    the list of ranges to scan
    * @param limit     number of result
    * @return a select query
    */
-  private PreparedStatement prepareMultiScanQuery(Map<String, Set<Field<?>>> keyFields,
+  private PreparedStatement prepareMultiScanQuery(Collection<Range> singletonRanges,
                                                   Collection<Range> ranges, int limit) throws SQLException {
```

Comment:
Generally it can be all collapsed into a single use case, as range can have few singleton fields and few non-singleton, that would be really nice to collapse into (key1 = ? and key2 >= ? and key2 <= ?). In this case singleton would just be a range will all singleton fields.... But it's outside of this PR scope.

Reply:
I am wondering what is perf effect of `(namespace >= 'default' and namespace <= 'default' and app >='a' and app <= 'b')` vs `(namespace = 'default' and app >='a' and app <= 'b')`. Probably negligible given index statistics

Reply:
I'd be surprised if sql cannot handle that efficiently

Reply:
Well, 
 * We use ancient PostgreSQL. Like decade old
 * In PostgreSQL community they prefer not to add logic that can be easily solved by query writer. They say it would take CPU cycles for every query run in the world
So, I won't be surprised that it won't collapse it to "=". It should still use proper statistics and an index scan though. "OR"s often make optmizer life complex as it's hard to decide between index scan and sequence scan (similarly to our PR here :) ), so having a simpler condition may help in some cases

Reply:
that could be a nice refactoring task, will add it